capsules/extra/src/ieee802154/driver.rs

```diff
@@ -1042,12 +1088,74 @@ impl<'a, M: device::MacDevice<'a>> device::RxClient for RadioDriver<'a, M> {
+                        // case of the buffer being shorter than the `RING_BUF_METADATA_SIZE` as an
+                        // invalid buffer (e.g. of length 1) may otherwise errantly pass the second
+                        // conditional check (due to unsigned integer arithmetic).
+                        if rbuf.len() <= RING_BUF_METADATA_SIZE
+                            || (rbuf.len() - RING_BUF_METADATA_SIZE) % USER_FRAME_MAX_SIZE != 0
+                        {
+                            // kernel::debug!("[15.4 Driver] Error - improperly formatted readwrite buffer provided");
+                            return false;
+                        }
+
+                        let mic_len = header.security.map_or(0, |sec| sec.level.mic_len());
+                        let frame_len = data_offset + data_len + mic_len;
+
+                        // Frame length incorporates the PSDU offset, so we must subtract this value
+                        // in addition to incorporating the user frame metadata size to obtain the
+                        // length of the user frame.
+                        let user_frame_len = frame_len + USER_FRAME_METADATA_SIZE - PSDU_OFFSET;
+
+                        let read_index = rbuf[0].get() as usize;
+                        let mut write_index = rbuf[1].get() as usize;
+
+                        let max_pending_rx =
+                            (rbuf.len() - RING_BUF_METADATA_SIZE) / USER_FRAME_MAX_SIZE;
+
+                        // confirm user modifiable metadata is valid (i.e. within bounds of the provided buffer)
+                        if read_index >= max_pending_rx || write_index >= max_pending_rx {
+                            // kernel::debug!("[15.4 driver] Invalid read or write index");
+                            return false;
+                        }
+
+                        let offset = RING_BUF_METADATA_SIZE + (write_index * USER_FRAME_MAX_SIZE);
+
+                        // Copy the entire frame over to userland, preceded by three metadata bytes:
+                        // the data offset, the data length, and the MIC length.
+                        rbuf[(offset + USER_FRAME_METADATA_SIZE)..(offset + user_frame_len)]
+                            .copy_from_slice(&buf[PSDU_OFFSET..frame_len]);
+                        rbuf[offset].set(data_offset as u8);
+                        rbuf[offset + 1].set(data_len as u8);
+                        rbuf[offset + 2].set(mic_len as u8);
+
+                        // Prepare the ring buffer for the next write. The current design favors newness;
+                        // newly received packets will begin to overwrite the oldest data in the event
+                        // of the buffer becoming full.
+                        write_index = (write_index + 1) % max_pending_rx;
+                        if write_index == read_index {
+                            // kernel::debug!("[15.4 driver] Provided RX buffer is full");
```

Comment:
Should we update `read_index` here? It might cause odd behavior if a new packet is "received" before an old one.